infra/Pos/Slotting/Types.hs

```diff
@@ -22,15 +38,109 @@ data EpochSlottingData = EpochSlottingData
+    } deriving (Eq, Show, Generic, Monoid)
 
 instance NFData SlottingData
+
+-- | Unsafe constructor that can lead to unsafe crash!
+createSlottingDataUnsafe :: Map EpochIndex EpochSlottingData -> SlottingData
+createSlottingDataUnsafe epochSlottingDataMap =
+    if M.size epochSlottingDataMap < 2
+        then criticalError
+        else SlottingData epochSlottingDataMap
+  where
+    criticalError = error "It's impossible to create slotting data without two epochs."
+
+-- | Restricted constructor function for the (initial) creation of @SlottingData@.
+createInitSlottingData
+    :: CurrentEpochSlottingData
+    -> NextEpochSlottingData
+    -> SlottingData
+createInitSlottingData psd esd = SlottingData validInitialSlottingData
+  where
+
+    validInitialSlottingData :: Map EpochIndex EpochSlottingData
+    validInitialSlottingData = M.union currentEpochSlottingData nextEpochSlottingData
+
+    currentEpochSlottingData :: Map EpochIndex CurrentEpochSlottingData
+    currentEpochSlottingData = M.singleton 0 psd
+
+    nextEpochSlottingData :: Map EpochIndex NextEpochSlottingData
+    nextEpochSlottingData = M.singleton 1 esd
+
+-- | Get all epoch index.
+getAllEpochIndex :: SlottingData -> [EpochIndex]
+getAllEpochIndex = M.keys . getSlottingDataMap
+
+-- | Get the next epoch index.
+getNextEpochIndex :: SlottingData -> EpochIndex
+getNextEpochIndex = fst . M.findMax . getSlottingDataMap
+
+-- | Get the next epoch slotting data.
+getNextEpochSlottingData :: SlottingData -> EpochSlottingData
+getNextEpochSlottingData = snd . M.findMax . getSlottingDataMap
+
+-- | Get the current epoch index. Next epoch - 1.
+getCurrentEpochIndex :: SlottingData -> EpochIndex
+getCurrentEpochIndex = decreaseEpochIndex . getNextEpochIndex
+  where
+    decreaseEpochIndex :: EpochIndex -> EpochIndex
+    decreaseEpochIndex ei = EpochIndex $ getEpochIndex ei - 1
+
+-- | Get the current epoch slotting data. Next epoch - 1.
+getCurrentEpochSlottingData :: SlottingData -> EpochSlottingData
+getCurrentEpochSlottingData sdp@(getSlottingDataMap -> sd) = sd M.! currentEpochIndex
+  where
+    currentEpochIndex = getCurrentEpochIndex sdp
+
+-- | Lookup the slotting data for an arbitrary `EpochIndex`.
+lookupEpochSlottingData :: EpochIndex -> SlottingData -> Maybe EpochSlottingData
+lookupEpochSlottingData epochIndex slottingData = M.lookup epochIndex slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Insert `EpochSlottingData`.
+addEpochSlottingData :: EpochIndex -> EpochSlottingData -> SlottingData -> SlottingData
+addEpochSlottingData epochIndex epochSlottingData slottingData =
+    SlottingData $ M.insert epochIndex epochSlottingData slottingData'
+  where
+    slottingData' :: Map EpochIndex EpochSlottingData
+    slottingData' = getSlottingDataMap slottingData
+
+-- | Compute when the slot started.
+computeSlotStart :: SlotId -> EpochSlottingData -> Timestamp
```

Comment:
Also, I think it would be much easier if we added a field `startTime` alongside with `startDiff` in `EpochSlottingData`.
Again, refactoring this would be a breeze if we could "hide" behind a typeclass - https://github.com/input-output-hk/cardano-sl/pull/1217#discussion_r130218756